pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,671 @@
+	ces, err := os.ReadDir(chunksDir)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read directory %s", chunksDir)
+	}
+	for _, e := range ces {
+		meta.Thanos.SegmentFiles = append(meta.Thanos.SegmentFiles, e.Name())
+		fi, err := analyzeBlockFile(blockDir, filepath.Join("chunks", e.Name()))
+		if err != nil {
+			return err
+		}
+
+		meta.Thanos.Files = append(meta.Thanos.Files, fi)
+	}
+
+	// Reset sources/parents since block history may be inconsistent with Mimir's
+	bULID, err := ulid.Parse(blockID)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse block ID")
+	}
+	// Sources should always contain block ID itself
+	meta.Compaction.Sources = []ulid.ULID{
+		bULID,
+	}
+	meta.Compaction.Parents = nil
+
+	if err := c.scanBlock(blockDir, meta); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func analyzeBlockFile(relPath, blockDir string) (metadata.File, error) {
+	fi, err := os.Stat(filepath.Join(blockDir, relPath))
+	if err != nil {
+		return metadata.File{}, errors.Wrapf(err, "failed to stat %s", filepath.Join(blockDir, relPath))
+	}
+	return metadata.File{
+		RelPath:   relPath,
+		SizeBytes: fi.Size(),
+	}, nil
+}
+
+func (c *MultitenantCompactor) scanBlock(blockDir string, meta metadata.Meta) error {
+	// TODO: Count samples (check with Peter)
+
+	var cr *chunks.Reader
+	cr, err := chunks.NewDirReader(filepath.Join(blockDir, block.ChunksDirname), nil)
+	if err != nil {
+		return errors.Wrap(err, "open chunks dir")
+	}
+	defer runutil.CloseWithErrCapture(&err, cr, "closing chunks reader")
+
+	r, err := index.NewFileReader(filepath.Join(blockDir, block.IndexFilename))
+	if err != nil {
+		return errors.Wrap(err, "open index file")
+	}
+	defer runutil.CloseWithErrCapture(&err, r, "gather index issue file reader")
+
+	ps, err := r.Postings(index.AllPostingsKey())
+	if err != nil {
+		return errors.Wrap(err, "get all postings")
+	}
+
+	shard, shardCount, err := sharding.ParseShardIDLabelValue(meta.Thanos.Labels[mimir_tsdb.CompactorShardIDExternalLabel])
+	if err != nil {
+		return errors.Wrap(err, "parse shard ID label value")
+	}
+
+	var (
+		lastLset labels.Labels
+		lset     labels.Labels
+
+		outOfOrderLabels int
+	)
+	// Per series.
+	for ps.Next() {
+		lastLset = append(lastLset[:0], lset...)
+
+		id := ps.At()
+
+		var chnks []chunks.Meta
+		if err := r.Series(id, &lset, &chnks); err != nil {
+			return errors.Wrap(err, "read series")
+		}
+		if len(lset) == 0 {
+			return errors.Errorf("empty label set detected for series %d", id)
+		}
+		if lastLset != nil && labels.Compare(lastLset, lset) >= 0 {
+			return errors.Errorf("series %v out of order; previous: %v", lset, lastLset)
+		}
+		l0 := lset[0]
+		for _, l := range lset[1:] {
+			if l.Name < l0.Name {
+				outOfOrderLabels++
+				level.Warn(c.logger).Log("msg",
+					"out-of-order label set: known bug in Prometheus 2.8.0 and below",
+					"labelset", lset.String(),
+					"series", fmt.Sprintf("%d", id),
+				)
+			}
+			l0 = l
+		}
+		if len(chnks) == 0 {
```

Comment:
Does this mean that we allow some out of order labels? I don't think we should do that.